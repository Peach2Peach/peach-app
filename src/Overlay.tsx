import { atom, useAtomValue, useSetAtom } from 'jotai'
import { Modal, View } from 'react-native'
import { useSafeAreaInsets } from 'react-native-safe-area-context'
import { PeachyBackground } from './components/PeachyBackground'
import { Popup } from './components/popup/Popup'
import tw from './styles/tailwind'

const overlayAtom = atom<React.ReactNode>(undefined)
export const useSetOverlay = () => useSetAtom(overlayAtom)

export function Overlay () {
  const content = useAtomValue(overlayAtom)
  const insets = useSafeAreaInsets()
  return (
    <Modal visible={content !== undefined}>
      <Popup />
      <PeachyBackground />
      <View style={{ paddingTop: insets.top, paddingBottom: insets.bottom, flex: 1 }}>
        <View style={[tw`flex-1 p-sm`, tw`md:p-md`]}>{content}</View>
      </View>
    </Modal>
  )
}
