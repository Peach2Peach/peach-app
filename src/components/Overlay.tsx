import { View } from 'react-native'
import { IconType } from '../assets/icons'
import tw from '../styles/tailwind'
import { Icon } from './Icon'
import { Screen } from './Screen'
import { PeachText } from './text/Text'

type Props = {
  buttons: React.ReactNode
  title: string
  text: string
  iconId?: IconType
}

export function Overlay ({ buttons, title, text, iconId }: Props) {
  return (
    <Screen gradientBackground>
      <View style={tw`grow`}>
        <View style={tw`justify-center gap-8 grow`}>
          <PeachText style={tw`text-center h4 text-primary-background-light shrink`}>{title}</PeachText>
          <View style={tw`flex-row items-center gap-6`}>
            {iconId && <Icon id={iconId} size={92} color={tw.color('primary-background-light')} />}
            <PeachText style={[tw`flex-1 body-l text-primary-background-light`, !iconId && tw`text-center`]}>
              {text}
            </PeachText>
          </View>
        </View>

        <View style={tw`self-center gap-3`}>{buttons}</View>
      </View>
    </Screen>
  )
}
