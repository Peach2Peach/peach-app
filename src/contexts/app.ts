import { createContext, Dispatch, ReducerState } from 'react'

let notifications = 0

export const getAppContext = (): AppState => ({
  notifications
})

const dispatch: Dispatch<AppState> = () => {}

export const AppContext = createContext([
  { notifications },
  dispatch
] as const)

export const setAppContext = (state: ReducerState<any>, newState: Partial<AppState>): AppState => {
  notifications = Math.max(0, newState.notifications ?? notifications)
  return {
    notifications
  }
}


export default AppContext