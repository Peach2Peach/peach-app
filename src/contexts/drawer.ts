import { createContext, Dispatch, ReactNode, ReducerState } from 'react'

let title = ''
let content: ReactNode
let show: boolean = false
let onClose = () => {}

const dispatch: Dispatch<Partial<DrawerState>> = () => {}

export const DrawerContext = createContext([{ title, content, show, onClose }, dispatch] as const)

/**
 * @description Method to get drawer content
 * @returns drawer content
 */
export const getDrawer = (): DrawerState => ({
  title,
  content,
  show,
  onClose,
})

/**
 * @description Method to set drawer content
 * @param state the state object (can be ignored)
 * @param newState the new state object
 * @returns drawer state
 */
export const setDrawer = (state: ReducerState<any>, newState: Partial<DrawerState>): DrawerState => {
  title = newState.title || title
  content = newState.content ?? content
  show = newState.show ?? true
  onClose = newState.onClose || onClose

  return {
    title,
    content,
    show,
    onClose,
  }
}
