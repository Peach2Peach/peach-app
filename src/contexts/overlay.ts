import { createContext, Dispatch, ReactNode, ReducerState } from 'react'

let content: ReactNode
let showCloseIcon = false
let showCloseButton = false
let help = false
let isTransparent = false

const dispatch: Dispatch<OverlayState> = () => {}

export const OverlayContext = createContext([
  { content, showCloseIcon, showCloseButton, help, isTransparent },
  dispatch
] as const)

/**
 * @description Method to get overlay content
 * @returns overlay content
 */
export const getOverlay = (): OverlayState => ({
  content,
  showCloseIcon,
  showCloseButton,
  help,
  isTransparent
})

/**
 * @description Method to set overlay content
 * @param state the state object (can be ignored)
 * @param newState the new state object
 * @returns overlay state
 */
export const setOverlay = (state: ReducerState<any>, newState: OverlayState): OverlayState => {
  content = newState.content
  showCloseIcon = newState.showCloseIcon ?? false
  showCloseButton = newState.showCloseButton ?? false
  help = newState.help ?? false
  isTransparent = newState.isTransparent ?? false

  return {
    content,
    showCloseIcon,
    showCloseButton,
    help,
    isTransparent
  }
}