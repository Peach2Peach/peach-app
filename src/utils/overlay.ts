import { createContext, Dispatch, ReactNode, ReducerState } from 'react'

let overlayContent: ReactNode

const dispatch: Dispatch<OverlayState> = () => {}

export const OverlayContext = createContext([
  { overlayContent },
  dispatch
] as const)

/**
 * @description Method to get overlay content
 * @returns overlay content
 */
export const getOverlay = (): OverlayState => ({
  overlayContent,
})

/**
 * @description Method to set overlay content
 * @param state the state object (can be ignored)
 * @param newState the new state object
 * @returns overlay state
 */
export const setOverlay = (state: ReducerState<any>, newState: OverlayState): OverlayState => {
  overlayContent = newState.overlayContent

  return {
    overlayContent,
  }
}