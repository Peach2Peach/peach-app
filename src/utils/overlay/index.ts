export { shouldShowBuyerCanceledTrade } from './shouldShowBuyerCanceledTrade'
export { shouldShowCancelTradeRequestConfirmed } from './shouldShowCancelTradeRequestConfirmed'
export { shouldShowCancelTradeRequestRejected } from './shouldShowCancelTradeRequestRejected'
export { shouldShowConfirmCancelTradeRequest } from './shouldShowConfirmCancelTradeRequest'
export { shouldShowDisputeResult } from './shouldShowDisputeResult'
export { shouldShowYouGotADispute } from './shouldShowYouGotADispute'
