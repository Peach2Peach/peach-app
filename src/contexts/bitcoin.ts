import { createContext } from 'react'
import { SATSINBTC } from '../constants'
import { marketPrice } from '../utils/peachAPI'

export let bitcoinContext: BitcoinContextType = {
  currency: 'EUR',
  price: NaN,
  satsPerUnit: NaN
}

export const getBitcoinContext = (): BitcoinContextType => bitcoinContext

/**
 * @description Context for bitcoin
 * @example
 * import BitcoinContext from './components/contexts/bitcoin'
 *
 * export default (): ReactElement =>
 *   const { locale } = useContext(BitcoinContext)
 *   return <Text>
 *     {locale}
 *   </Text>
 * }
 */
export const BitcoinContext = createContext(bitcoinContext)


export const updateBitcoinContext = async (currency: Currency): Promise<BitcoinContextType> => {
  let price = bitcoinContext.price
  let satsPerUnit = bitcoinContext.satsPerUnit

  const [pairInfo, error] = await marketPrice(currency)

  if (!pairInfo || error?.error) {
    return bitcoinContext
  }

  price = Number(pairInfo.price)
  satsPerUnit = Math.round(SATSINBTC / price)
  // eslint-disable-next-line require-atomic-updates
  bitcoinContext = {
    currency,
    price,
    satsPerUnit
  }

  return bitcoinContext
}

export const bitcoinContextEffect = (
  context: BitcoinContextType,
  setBitcoinContext: React.Dispatch<React.SetStateAction<BitcoinContextType>>
) => () => {
  let interval: NodeJS.Timer

  const checkingFunction = async () => {
    // TODO add error handling in case data is not available
    setBitcoinContext(await updateBitcoinContext(context.currency))
  }
  (async () => {
    interval = setInterval(checkingFunction, 30 * 1000)
    checkingFunction()
  })()
  return () => {
    clearInterval(interval)
  }
}

export default BitcoinContext