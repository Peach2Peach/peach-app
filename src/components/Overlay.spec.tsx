import { Overlay } from './Overlay'
import { createRenderer } from 'react-test-renderer/shallow'
import { act, fireEvent, render } from '@testing-library/react-native'
import { Text, BackHandler } from 'react-native'
import { defaultOverlay, OverlayContext } from '../contexts/overlay'

jest.mock('react-native/Libraries/Utilities/BackHandler', () =>
  jest.requireActual('react-native/Libraries/Utilities/__mocks__/BackHandler.js'),
)

describe('Overlay', () => {
  let overlayState = { ...defaultOverlay, visible: true }
  const updateOverlay = jest.fn((newOverlayState: Partial<OverlayState>) => {
    overlayState = {
      ...overlayState,
      ...newOverlayState,
    }
  })
  const wrapper = ({ children }: { children: JSX.Element }) => (
    <OverlayContext.Provider value={[defaultOverlay, updateOverlay]}>{children}</OverlayContext.Provider>
  )
  const renderer = createRenderer()

  beforeEach(() => {
    overlayState = { ...defaultOverlay, visible: true }
    jest.clearAllMocks()
  })
  it('should render correctly', () => {
    renderer.render(<Overlay {...overlayState} level={undefined} />)
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
  it('should render correctly with existing, disabled actions and "WARN" level', () => {
    renderer.render(
      <Overlay
        {...overlayState}
        level="WARN"
        action1={{ label: 'action1', disabled: true, icon: 'testIcon', callback: jest.fn() }}
        action2={{ label: 'action2', disabled: true, callback: jest.fn() }}
      />,
    )
    const result = renderer.getRenderOutput()
    expect(result).toMatchSnapshot()
  })
  it('should call action1 callback when action1 is pressed', () => {
    const action1Callback = jest.fn()
    const { getByText } = render(
      <Overlay
        {...overlayState}
        action1={{ label: 'action1', callback: action1Callback }}
        action2={{ label: 'action2', callback: jest.fn() }}
      />,
      { wrapper },
    )
    fireEvent.press(getByText('action1'))
    expect(action1Callback).toHaveBeenCalled()
  })
  it('should not call action1 callback when action1 is pressed and action1 is disabled', () => {
    const action1Callback = jest.fn()
    const { getByText } = render(
      <Overlay
        {...overlayState}
        action1={{ label: 'action1', disabled: true, callback: action1Callback }}
        action2={{ label: 'action2', callback: jest.fn() }}
      />,
      { wrapper },
    )
    fireEvent.press(getByText('action1'))
    expect(action1Callback).not.toHaveBeenCalled()
  })
  it('should close the overlay when the hardware back button is pressed', () => {
    render(<Overlay {...overlayState} content={<Text>TestContent</Text>} />, { wrapper })

    act(() => {
      // @ts-ignore
      BackHandler.mockPressBack()
    })
    expect(overlayState.visible).toBe(false)
  })
  it('should not close the overlay when the hardware back button is pressed and there is no content', () => {
    render(<Overlay {...overlayState} />, { wrapper })

    act(() => {
      // @ts-ignore
      BackHandler.mockPressBack()
    })
    expect(overlayState.visible).toBe(true)
  })
  it('should close the overlay when the background is pressed', () => {
    const { getByTestId } = render(<Overlay {...overlayState} />, { wrapper })

    act(() => {
      fireEvent.press(getByTestId('overlay-background'))
    })
    expect(overlayState.visible).toBe(false)
  })
  it('should not close the overlay when the background is pressed and requireUserAction is true', () => {
    const { getByTestId } = render(<Overlay {...overlayState} requireUserAction={true} />, { wrapper })

    act(() => {
      fireEvent.press(getByTestId('overlay-background'))
    })
    expect(overlayState.visible).toBe(true)
  })
})
