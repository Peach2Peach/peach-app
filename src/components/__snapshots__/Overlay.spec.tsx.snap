// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overlay should render correctly 1`] = `
<PopupContent
  closePopup={[Function]}
  level="DEFAULT"
  requireUserAction={false}
  title=""
  visible={true}
/>
`;

exports[`Overlay should render correctly with existing, disabled actions and "WARN" level 1`] = `
<PopupContent
  action1={
    {
      "callback": [MockFunction],
      "disabled": true,
      "icon": "testIcon",
      "label": "action1",
    }
  }
  action2={
    {
      "callback": [MockFunction],
      "disabled": true,
      "label": "action2",
    }
  }
  closePopup={[Function]}
  level="WARN"
  requireUserAction={false}
  title=""
  visible={true}
/>
`;
